src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -566,11 +566,23 @@ private function validateAndDenormalizeLegacy(array $types, string $currentClass
                     return (float) $data;
                 }
 
-                if ((LegacyType::BUILTIN_TYPE_FALSE === $builtinType && false === $data) || (LegacyType::BUILTIN_TYPE_TRUE === $builtinType && true === $data)) {
-                    return $data;
-                }
-
-                if (('is_'.$builtinType)($data)) {
+                $dataMatchesExpectedType = match ($builtinType) {
+                    LegacyType::BUILTIN_TYPE_ARRAY => \is_array($data),
+                    LegacyType::BUILTIN_TYPE_BOOL => \is_bool($data),
+                    LegacyType::BUILTIN_TYPE_CALLABLE => \is_callable($data),
+                    LegacyType::BUILTIN_TYPE_FALSE => false === $data,
+                    LegacyType::BUILTIN_TYPE_FLOAT => \is_float($data),
+                    LegacyType::BUILTIN_TYPE_INT => \is_int($data),
+                    LegacyType::BUILTIN_TYPE_ITERABLE => is_iterable($data),
+                    LegacyType::BUILTIN_TYPE_NULL => null === $data,
+                    LegacyType::BUILTIN_TYPE_OBJECT => \is_object($data),
+                    LegacyType::BUILTIN_TYPE_RESOURCE => $data,
+                    LegacyType::BUILTIN_TYPE_STRING => \is_string($data),
+                    LegacyType::BUILTIN_TYPE_TRUE => true === $data,
+                    default => false,
```

Comment:
In the linked bug report, people explicitly mentioned the `mixed` type. I don't see it handled here and `false` as a fallback does not seem accurate: Anything can be mapped to mixed, so `$dataMatchesExpectedType` should be always true if we encounter `mixed`, shouldn't it?

Reply:
Yes, you are right. I was able to extract a test case and adding what you suggested solves the issue.